src/Compilers/Shared/BuildServerConnection.cs

```diff
@@ -595,4 +612,216 @@ public static string GetTempPath(string workingDir)
             return Environment.GetEnvironmentVariable("SYSTEMROOT");
         }
     }
+
+    internal interface IServerMutex : IDisposable
+    {
+        bool TryLock(int timeoutMs);
+        void Unlock();
+        bool IsDisposed { get; }
+    }
+
+    internal sealed class FileMutex : IDisposable
```

Comment:
Should probably add a doc comment explaining how this lock works and what the basic invariants are.